vpn/version.go

```diff
@@ -1,10 +1,138 @@
+// not backwards compatible with RPCVersion{1, 2}.
+type RPCVersion struct {
+	Major uint64 `json:"major"`
+	Minor uint64 `json:"minor"`
+}
+
+// ParseRPCVersion parses a version string in the format "major.minor" into a
+// RPCVersion.
+func ParseRPCVersion(str string) (RPCVersion, error) {
+	split := strings.Split(str, ".")
+	if len(split) != 2 {
+		return RPCVersion{}, xerrors.Errorf("invalid version string: %s", str)
+	}
+	major, err := strconv.ParseUint(split[0], 10, 64)
+	if err != nil {
+		return RPCVersion{}, xerrors.Errorf("invalid version string: %s", str)
+	}
+	if major == 0 {
+		return RPCVersion{}, xerrors.Errorf("invalid version string: %s", str)
+	}
+	minor, err := strconv.ParseUint(split[1], 10, 64)
+	if err != nil {
+		return RPCVersion{}, xerrors.Errorf("invalid version string: %s", str)
+	}
+	return RPCVersion{Major: major, Minor: minor}, nil
+}
+
+func (v RPCVersion) String() string {
+	return fmt.Sprintf("%d.%d", v.Major, v.Minor)
+}
+
+// IsCompatibleWith returns the lowest version that is compatible with both
+// versions. If the versions are not compatible, the second return value will be
+// false.
+func (v RPCVersion) IsCompatibleWith(other RPCVersion) (RPCVersion, bool) {
+	if v.Major != other.Major {
+		return RPCVersion{}, false
+	}
+	// The lowest minor version from the two versions should be returned.
+	if v.Minor < other.Minor {
+		return v, true
+	}
+	return other, true
+}
+
+// RPCVersionList represents a list of RPC versions supported by a RPC peer. An
+type RPCVersionList struct {
+	Versions []RPCVersion `json:"versions"`
+}
+
+// ParseRPCVersionList parses a version string in the format
+// "major.minor,major.minor" into a RPCVersionList.
+func ParseRPCVersionList(str string) (RPCVersionList, error) {
+	split := strings.Split(str, ",")
+	versions := make([]RPCVersion, len(split))
+	for i, v := range split {
+		version, err := ParseRPCVersion(v)
+		if err != nil {
+			return RPCVersionList{}, xerrors.Errorf("invalid version string: %s", v)
+		}
+		versions[i] = version
+	}
+	vl := RPCVersionList{Versions: versions}
+	err := vl.Validate()
+	if err != nil {
+		return RPCVersionList{}, xerrors.Errorf("invalid parsed version list %q: %w", str, err)
+	}
+	return vl, nil
+}
+
+func (vl RPCVersionList) String() string {
+	versionStrings := make([]string, len(vl.Versions))
+	for i, v := range vl.Versions {
+		versionStrings[i] = v.String()
+	}
+	return strings.Join(versionStrings, ",")
+}
+
+// Validate returns an error if the version list is not sorted or contains
+// duplicate major versions.
+func (vl RPCVersionList) Validate() error {
+	if len(vl.Versions) == 0 {
+		return xerrors.New("no versions")
+	}
+	for i := 1; i < len(vl.Versions); i++ {
+		if vl.Versions[i-1].Major == vl.Versions[i].Major {
+			return xerrors.Errorf("duplicate major version: %d", vl.Versions[i].Major)
+		}
+		if vl.Versions[i-1].Major > vl.Versions[i].Major {
+			return xerrors.Errorf("versions are not sorted")
```

Comment:
Why does the list need to be sorted? Should it sorted by both major and minor?

Reply:
No, because we don't allow duplicate major versions.

> The versions list is a list of the maximum supported minor version for each major version, sorted by major versions.

Reply:
Okay, yeah, I'm dumb